EclipseLink for JPA 3.1 may encounter IllegalArgumentException: Unsupported api 0

**Describe the bug**  
When using EclipseLink from the JPA 3.1 feature, the following error may occur if `org.eclipse.persistence.asm.Opcodes` is initialized before `org.eclipse.persistence.asm.ASMFactory`:
```
Caused by: java.lang.IllegalArgumentException: Unsupported api 0
	at org.eclipse.persistence.internal.libraries.asm.ClassVisitor.<init>(ClassVisitor.java:77)
	at org.eclipse.persistence.asm.internal.platform.eclipselink.ClassVisitorImpl$ElClassVisitor.<init>(ClassVisitorImpl.java:27)
	at org.eclipse.persistence.asm.internal.platform.eclipselink.ClassVisitorImpl.<init>(ClassVisitorImpl.java:106)
	at org.eclipse.persistence.asm.ASMFactory.createClassVisitor(ASMFactory.java:194)
	at org.eclipse.persistence.asm.ClassVisitor.<init>(ClassVisitor.java:26)
	at org.eclipse.persistence.asm.EclipseLinkClassVisitor.<init>(EclipseLinkClassVisitor.java:20)
	at org.eclipse.persistence.internal.jpa.metadata.accessors.objects.MetadataAsmFactory$ClassMetadataVisitor.<init>(MetadataAsmFactory.java:269)
	at org.eclipse.persistence.internal.jpa.metadata.accessors.objects.MetadataAsmFactory.buildClassMetadata(MetadataAsmFactory.java:91)
	at org.eclipse.persistence.internal.jpa.metadata.accessors.objects.MetadataAsmFactory.getMetadataClass(MetadataAsmFactory.java:184)
	at org.eclipse.persistence.internal.jpa.metadata.ORMetadata.getMetadataClass(ORMetadata.java:408)
	at org.eclipse.persistence.internal.jpa.metadata.xml.XMLEntityMappings.initPersistenceUnitClasses(XMLEntityMappings.java:642)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataProcessor.initPersistenceUnitClasses(MetadataProcessor.java:299)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataProcessor.processEntityMappings(MetadataProcessor.java:583)
	at org.eclipse.persistence.internal.jpa.deployment.PersistenceUnitProcessor.processORMetadata(PersistenceUnitProcessor.java:640)
```

**Expected behavior**  
The IllegalArgumentException should not occur; the metadata should be processed without error.

**Diagnostic information:**  
 - OpenLiberty Version: 23.0.0.7
 - Affected feature(s) : jpa-3.1
 - Java Version: any
 
**Additional context**  
None

--- a/dev/io.openliberty.persistence.3.1.thirdparty/src/org/eclipse/persistence/asm/ASMFactory.java
+++ b/dev/io.openliberty.persistence.3.1.thirdparty/src/org/eclipse/persistence/asm/ASMFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023,2024 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0 which is available at

--- a/dev/io.openliberty.persistence.3.1.thirdparty/src/org/eclipse/persistence/asm/ASMFactory.java
+++ b/dev/io.openliberty.persistence.3.1.thirdparty/src/org/eclipse/persistence/asm/ASMFactory.java
@@ -14,6 +14,8 @@
 //     Oracle - initial API and implementation
 package org.eclipse.persistence.asm;
 
+import org.eclipse.persistence.asm.internal.Util;
+
 import org.eclipse.persistence.config.SystemProperties;
 import org.eclipse.persistence.exceptions.ValidationException;
 import org.eclipse.persistence.internal.security.PrivilegedAccessHelper;
@@ -26,6 +28,7 @@ import java.lang.reflect.Field;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
@@ -33,17 +36,36 @@ import java.util.regex.Pattern;
 
 public class ASMFactory {
 
-    private final static SessionLog LOG = AbstractSessionLog.getLog();
-
+    // This block must be first - begin
     public final static String ASM_SERVICE_ECLIPSELINK = "eclipselink";
     public final static String ASM_SERVICE_OW2 = "ow2";
+
     private final static String ASM_OW2_CLASS_VISITOR = "org.objectweb.asm.ClassVisitor";
     private final static String ASM_ECLIPSELINK_CLASS_VISITOR = "org.eclipse.persistence.internal.libraries.asm.ClassVisitor";
 
-    //Should be changed in case of ASM upgrade
-    public final static int ASM_API_SELECTED = Opcodes.ASM9;
-    public final static int JAVA_CLASS_VERSION = Opcodes.V1_8;
+    private final static String ASM_OPCCODES_OW2 = "org.objectweb.asm.Opcodes";
+    private final static String ASM_OPCCODES_ECLIPSELINK = "org.eclipse.persistence.internal.libraries.asm.Opcodes";
+
+    private final static Map<String, String> ASM_OPCCODES_MAP = new HashMap<>();
+
+    static {
+        ASM_OPCCODES_MAP.put(ASMFactory.ASM_SERVICE_OW2, ASM_OPCCODES_OW2);
+        ASM_OPCCODES_MAP.put(ASMFactory.ASM_SERVICE_ECLIPSELINK, ASM_OPCCODES_ECLIPSELINK);
+    }
+
+    private final static SessionLog LOG = AbstractSessionLog.getLog();
+    // This block must be first - end
+
+    // Do not reference static fields in Opcodes to avoid circular static initialization
+
+    // Should be changed in case of ASM upgrade
+    public final static int ASM_API_SELECTED = valueOpcodesInt("ASM9");
+    public final static int JAVA_CLASS_VERSION = valueOpcodesInt("V1_8");
     public final static int JAVA_CLASS_LATEST_VERSION = ASMFactory.getLatestOPCodeVersion();
+
+    private static int valueOpcodesInt(String fieldName) {
+        return ((int) Util.getFieldValue(ASM_OPCCODES_MAP, fieldName, Integer.TYPE));
+    }
 
     public static AnnotationVisitor createAnnotationVisitor(final int api) {
         String asmService = ASMFactory.getAsmService();

